### DISCUSSION TITLE
DOCS BUG: Docs for WixToolset.Dtf.CustomAction are missing

### DISCUSSION THREAD

**Discussion by BryanCrotaz** [wixtoolset/issues#7831] [Questions]
https://wixtoolset.org/docs/api/

Docs for all the Nuget libraries except CustomAction.

Following the examples (https://www.advancedinstaller.com/versus/wix-toolset/create-wix-custom-dialog-user-input.html), I installed WixToolset.Dtf.CustomAction. I've tried framework 2.0, 3.5 and 4.8 but every time I get the below errors. I'd like to look at the documentation to see how to use it correctly, but there is none.

```csharp
using WixToolset.Dtf.CustomAction;

namespace DownloadLicence
{
	public class CustomActions
	{
		[CustomAction]
		public static ActionResult DownloadLicence(Session session)
		{
			string tenantKey = session["TENANT_KEY"];
			// do stuff
			return ActionResult.Success;
		}
	}
}
```

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	CS0234	The type or namespace name 'CustomAction' does not exist in the namespace 'WixToolset.Dtf' (are you missing an assembly reference?)	DownloadLicence	H:\Projects\Client2\Client\DownloadLicence\CustomActions.cs	1	Active
Error	CS0246	The type or namespace name 'CustomActionAttribute' could not be found (are you missing a using directive or an assembly reference?)	DownloadLicence	H:\Projects\Client2\Client\DownloadLicence\CustomActions.cs	7	Active
Error	CS0246	The type or namespace name 'CustomAction' could not be found (are you missing a using directive or an assembly reference?)	DownloadLicence	H:\Projects\Client2\Client\DownloadLicence\CustomActions.cs	7	Active
Error	CS0246	The type or namespace name 'ActionResult' could not be found (are you missing a using directive or an assembly reference?)	DownloadLicence	H:\Projects\Client2\Client\DownloadLicence\CustomActions.cs	8	Active
Error	CS0246	The type or namespace name 'Session' could not be found (are you missing a using directive or an assembly reference?)	DownloadLicence	H:\Projects\Client2\Client\DownloadLicence\CustomActions.cs	8	Active
```

Nuget says your package is installed, but it's not listed under dependencies. Is there a prerequisite it needs? Hopefully the docs can tell me.  Oh.

**Comment by chrpai** [comment]
You really should give me an hour of your time to help you. I have 26 years experience creating installers for windows, 20 years experience with Windows Installer and 17 years experience with WiX.  I know this domain like the back of my hand and can help you avoid all the problems you are having.

**Reply by BryanCrotaz** [reply]
I don't have any budget for consulting I'm afraid. Firegiant's hourly rate is my daily rate! This is a bootstrapped project. 

I've got enough working that I have an installer, albeit not as good as I wanted.

I would love to contribute more to the docs but it's still magic incantation level to me so I don't know how to write something for a beginner. 

I had already found isWix but although I found the samples, I couldn't find the tutorials. The samples were very complicated with lots of detail I didn't understand. I think that's where I got my ui code from in the end but it was random "change this and see what breaks" rather than "oh I see how it works".

**Reply by chrpai** [reply]
This is the last time I will offer.  Please reach out to me.  I will give you an hour of my time for free. I can probably rewrite your installer from scratch and make it perfect during that time.

**Reply by BryanCrotaz** [reply]
If it's free then how can I turn you down? In return, if it will be accepted by @robmen I'd be happy to write a tutorial in wixtoolset of what I've learned from it for beginners

**Reply by robmen** [reply]
@BryanCrotaz, I will be honest with you. All of my interactions with you to this point have been negative. For that reason, I am not interested in reviewing and maintaining content that you produce. You will need to find another WiX maintainer to interact with to get pull requests accepted here.

Sadly, I do recognize that if you choose to write your beginner's tutorial, it will likely have to be on your website, and that will perpetuate the dispersion of documentation across the Internet.

But, on the upside, your behavior here has made me question several assumptions I've held for a long time, and I will be making changes.

May you find more interested collaborators in the future.

**Reply by BryanCrotaz** [reply]
@robmen a tip for the future then. When someone makes the effort to improve the product that's frustrating the heck out of them, and pushes a pull request...

If you don't like the code, don't just shut them down with a half sentence, close the PR and worst, lock the comments. That last bit is the piece that pissed me off more than anything else. That's the bit that said "I don't care that you want to help improve this, your contribution wasn't perfect, and I'm not interested in telling you why or helping you understand how to make it acceptable, just go away and stop bothering me" - one of them was a misreading on your part and I could see what had happened but I couldn't tell you because you'd locked it. That's where my head hit the desk and I realised I probably couldn't deliver to my customer because nothing was working, I couldn't grok the docs, I wasn't the only one - stack overflow is full of people having the same problem, and I couldn't even help fix the problem.

**Reply by robmen** [reply]
I don't lock comments. The CLA bot locks comments on PRs for legal purposes. That is definitely not my favorite behavior. As I understand it, the lawyers recommend it so that once someone accepts the CLA they cannot remove their acceptance after the PR is closed.

**Reply by BryanCrotaz** [reply]
Then maybe don't be in such a hurry to close PRs?
